stan/math/rev/fun/svd_U.hpp

```diff
@@ -0,0 +1,80 @@
+#ifndef STAN_MATH_REV_FUN_SVD_U_HPP
+#define STAN_MATH_REV_FUN_SVD_U_HPP
+
+#include <stan/math/rev/meta.hpp>
+#include <stan/math/rev/core.hpp>
+#include <stan/math/rev/fun/value_of.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/err/check_symmetric.hpp>
+#include <stan/math/prim/err/check_nonzero_size.hpp>
+#include <stan/math/prim/fun/typedefs.hpp>
+#include <stan/math/prim/fun/value_of_rec.hpp>
+#include <Eigen/SVD>
+
+namespace stan {
+namespace math {
+
+/**
+ * Given input matrix m, return matrix U where `m = UDV^{T}`
+ *
+ * This function internally calls Eigen::JacobiSVD to compute decomposition.
+ *
+ * Equation (4) from Differentiable Programming Tensor Networks(H. Liao, J. Liu,
+ * et al., arXiv:1903.09650) is used for MxN input matrix's adjoint calculation.
+ *
+ * @tparam EigMat type of input matrix
+ * @param m MxN input matrix
+ * @return Orthogonal matrix U
+ */
+template <typename EigMat, require_eigen_matrix_dynamic_t<EigMat>* = nullptr,
+          require_vt_var<EigMat>* = nullptr>
+inline auto svd_U(const EigMat& m) {
+  using ret_type = promote_scalar_t<var, Eigen::MatrixXd>;
+  check_nonzero_size("svd_U", "m", m);
+
+  const int M = to_arena(std::min(m.rows(), m.cols()));
+  auto arena_m = to_arena(m);  // N by P
+
+  Eigen::JacobiSVD<Eigen::MatrixXd> svd(
+      arena_m.val(), Eigen::ComputeThinU | Eigen::ComputeThinV);
+
+  auto arena_D = to_arena(svd.singularValues());  // size min(N, P) = M
+
+  Eigen::MatrixXd Fp(M, M);
```

Comment:
Do `arena_t<Eigen:MatrixXd> arena_Fp(M, M);`

And then just assign stuff directly there and you don't have to copy it later.